File: Tree/623.Add-One-Row-to-Tree/623.Add-One-Row-to-Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode addOneRow(TreeNode root, int val, int depth) {
        return addOneRowHelper(root, val, depth, 1);
    }

    private TreeNode addOneRowHelper(TreeNode root, int val, int depth, int direction) {
        if (depth == 1) {
            return direction == 1 ? new TreeNode(val, root, null) : new TreeNode(val, null, root);
        }

        if (root.left != null || depth == 2) {
            root.left = addOneRowHelper(root.left, val, depth - 1, 1);
        }

        if (root.right != null || depth == 2) {
            root.right = addOneRowHelper(root.right, val, depth - 1, 0);
        }

        return root;
    }
}

File: Tree/623.Add-One-Row-to-Tree/623.Add-One-Row-to-Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def addOneRow(self, root: Optional[TreeNode], val: int, depth: int) -> Optional[TreeNode]:
        def dfs(root, depth, direction):
            # 可能是最后一层
            if depth == 1:
                if direction == 1:
                    return TreeNode(val, root)
                else:
                    return TreeNode(val, None, root)
            if not root:
                return root
            root.left = dfs(root.left, depth - 1, 1)
            root.right = dfs(root.right, depth - 1, -1)
            return root
        return dfs(root, depth, 1) #可能在第一层
                
